### DISCUSSION TITLE
configuring the compile command string.

### DISCUSSION THREAD

**Discussion by Littlechay** [orbitalquark/textadept#550] [Q&A] [answered]
I've just discovered textadept and like what I see. I'm try to configure it for a bit of C development for microcontrollers. 

My compile command as used in Geany etc.. is:
`cd build && rm -rf * && cmake ..`

my build command is:
`cd build && make`

and then I want to configure a run command like:
 `sudo -S openocd -f interface/cmsis-dap.cfg -f target/rp2040.cfg -c "adapter speed 5000" -c "program %d/build/%e.elf verify reset exit`

However I have fallen at the first hurdle :)

I attempted to configure the compile command in init.lua like this:
`textadept.run.compile_commands.c = 'cd build && rm -rf * && cmake ..'`

The command appears when I select compile from the menu but when I hit ENTER to run it I get an error that cd failed to run no such file or directory. The output buffer looks like this, the paths are all good:

```
> cd /home/christopher/picoDev/isquaredc
> cd build && rm -rf * && cmake ..
/home/christopher/textadept/modules/textadept/run.lua:150: cd build && rm -rf * && cmake ..: Failed to execute child process “cd” (No such file or directory)
```

I know SFA about lua but am keen to learn .. Any suggestions on how to set this up more than welcome.

OH... OS is Debian 12

Cheers
Chris

**Comment by orbitalquark** [comment]
Textadept's compile/run/build/test/project commands are not run directly in a shell like bash. Instead, they are treated as executables with arguments and invoked directly. For your first case, you'd either create a shell script and have Textadept invoke that, or you'd simply run `cmake` and accept the incremental build result.

For your second case, `make -C build` will work.

I'm unsure about your third use case. That trailing `exit` looks problematic, but I don't know if that's an argument or a shell command. When it comes to interactive prompts, Textadept sends the entire line to the process, so you'd also want to pass `-p ''` or `--prompt ''` to `sudo`, and then enter your password. Keep in mind Textadept will not mask any passwords if a command is waiting for input.

Hopefully this helps a bit.

**Reply by Littlechay** [reply]
Hi thanks for that. Yes it help :) One step at a time!

This works `textadept.run.compile_commands.c = 'cmake %d/build'
` for compile.

I have this in init.lua for build `textadept.run.build_commands = 'make'
` but when I select build from the tools menu... absolutely nothing happens. I guess I need to somehow associate it with *.c files ... How do I do that? 

This is what I have in the init.lua file for these commands
```
-- my stuff
textadept.run.compile_commands.c = 'cmake %d/build'
textadept.run.build_commands = 'make'
```

Working on stage three now :) 

Edit: I also tried your option two in init.lua like this `textadept.run.build_commands = 'make -C build'` same thing it doesn't seem to be picked up by selecting build from the tools menu or shift+ctrl+b.

**Reply by Littlechay** [reply]
OK; so I got my third case above (send the file to the programmer) working by using textadept to run a bash script.  My script is like this:
```
#!/usr/bin/bash
echo [password] | sudo -S openocd -f interface/cmsis-dap.cfg -f target/rp2040.cfg -c "adapter speed 5000" -c "program $1/build/$2.elf verify reset exit"
```

And using this in the init.lua to call the script with the run command.
`textadept.run.run_commands.c = '%d/picotool.sh %d %e'`

I'm still stuck on the build command though. I must have screwed up something fundamental or am missing a really basic step as selecting build from the tools menu or Ctrl+shift+B appears to do absolutely nothing... in other words no command appears above the status bar (not even an empty command bar), nothing in the output buffer.

Any pointers on troubleshooting this please?

Solve this build issue and I have a fully functioning raspberry pi pico microcontroller IDE that is beautifully lightweight, slick and elegant.. IMNSHO :)

**Reply by orbitalquark** [reply]
`textadept.run.build_commands` is a table of project directories assigned to build commands. (You cannot assign a bare string like you did.) The project directories must be under version control (or at least have a top-level *.git/*, *.hg/*, etc. directory). Whenever you run Tools > Build, Textadept looks at the current file, finds its project, and then invokes that project's build command.

Normally, if your `Makefile` is in the project root, then you don't have to do anything and Textadept will simply invoke that. However, since yours is in *build/*, you need to set `textadept.run.build_commands['/path/to/project/root'] = 'make -C build'`.

**Reply by Littlechay** [reply]
> The project directories must be under version control (or at least have a top-level _.git/_, _.hg/_, etc. directory). Whenever you run Tools > Build, Textadept looks at the current file, finds its project, and then invokes that project's build command.
> 

This is a new concept for me, needing a project directory with specific naming conventions... Can you give me an example of the structure that textadept would be satisfied with?  Currently the build command doesn't trigger even if I set the '/path/to/project/root'  e.g. '%d/build' .

**Reply by orbitalquark** [reply, marked as answer]
Let's suppose you have a project at */home/me/projects/myproject*. In order for Textadept to recognize it as a project, there needs to be a version control folder in the root directory of that project. You have a couple of options:

1. If you have Git or Mercurial installed, you can open a terminal and run `git init` or `hg init` from the *myproject* directory.
2. You can create an empty *.git* or *.hg* directory in the *myproject* directory. Note that this is a "hidden" folder and may not be visible by default in your file browser or terminal. You can use a file browser or terminal to do this.

Since your *Makefile* is in */home/me/projects/myproject/build*, you'd tell Textadept about it like this:

`textadept.run.build_commands['/home/me/projects/myproject'] = 'make -C build'`

Note that the '%' macros for compile and run commands do not work for build, test, or project commands.

**Reply by Littlechay** [reply]
Awesome, thanks, that the was the missing piece of the puzzle for me. I'll write this up and share it later in case anybody else is following my path :)
